runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -90,4 +113,161 @@ public void healNode(@Nonnull String endpointToHeal, int retry, @Nonnull Duratio
+        if (responsiveLayoutServers.size() != layout.getLayoutServers().size()) {
+            clusterStatus = ClusterStatus.DEGRADED;
+            int quorumSize = (layout.getLayoutServers().size() / 2) + 1;
+            if (responsiveLayoutServers.size() < quorumSize) {
+                clusterStatus = ClusterStatus.UNAVAILABLE;
+            }
+        }
+
+        // The primary sequencer should be reachable for the cluster to be STABLE.
+        if (!peerResponsiveNodes.contains(layout.getPrimarySequencer())) {
+            clusterStatus = ClusterStatus.UNAVAILABLE;
+        }
+
+        // Check the availability of the log servers in all segments as reads to all addresses
+        // should be accessible.
+        for (LayoutSegment layoutSegment : layout.getSegments()) {
+
+            if (layoutSegment.getReplicationMode().equals(ReplicationMode.CHAIN_REPLICATION)) {
+                // Every node is required to be reachable in CHAIN_REPLICATION.
+                if (!peerResponsiveNodes.containsAll(layoutSegment.getAllLogServers())) {
+                    clusterStatus = ClusterStatus.UNAVAILABLE;
+                }
+            } else if (layoutSegment.getReplicationMode()
+                    .equals(ReplicationMode.QUORUM_REPLICATION)) {
+                // At least a quorum of nodes should be reachable in every stripe for the cluster
+                // to be STABLE.
+                for (LayoutStripe layoutStripe : layoutSegment.getStripes()) {
+                    List<String> responsiveLogServers
+                            = new ArrayList<>(layoutStripe.getLogServers());
+                    // Retain only the responsive servers.
+                    responsiveLogServers.retainAll(peerResponsiveNodes);
+
+                    if (!responsiveLogServers.containsAll(layoutStripe.getLogServers())) {
+                        if (clusterStatus.equals(ClusterStatus.STABLE)) {
+                            clusterStatus = ClusterStatus.DEGRADED;
+                        }
+                        int quorumSize = (layoutStripe.getLogServers().size() / 2) + 1;
+                        if (responsiveLogServers.size() < quorumSize) {
+                            clusterStatus = ClusterStatus.UNAVAILABLE;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return clusterStatus;
+    }
+
+    /**
+     * Gets the cluster status by pinging each node.
+     * These pings are then compared to the layout to decide whether the cluster is:
+     * STABLE, DEGRADED OR UNAVAILABLE.
+     * The report consists of the following:
+     * Layout - at which the report was generated.
+     * ClientServerConnectivityStatusMap - the connectivity status of this client to the cluster.
+     * ClusterStatus - Health of the Corfu cluster.
+     *
+     * @return ClusterStatusReport
+     */
+    public ClusterStatusReport getClusterStatus() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)